### ISSUE TITLE
Implementation of DimeNet++

### ISSUE THREAD

**Issue by kryczko** [rusty1s/pytorch_geometric#2403]
From now on, we recommend using our discussion forum (https://github.com/rusty1s/pytorch_geometric/discussions) for proposing new features.

## 🚀 Feature
A implementation of the DimeNet++ model. 
## Motivation
The DimeNet++ model has shown to have better accuracy and is more efficient than it's V1 counterpart.

## Additional context
The DimeNet++ pretrained model is already being hosted on Github.

**Comment by thegodone** [comment]
Cautions guys there are already 3 bugs reported in the original tensorflow code. I would suggest to fix them first. I have DimeNet++ model with PyG I was able to fix one bug but not the 2 others.

**Comment by rusty1s** [comment]
There are plans to integrate DimeNet++. @thegodone Which bugs are you referring to?

**Comment by thegodone** [comment]
![image](https://user-images.githubusercontent.com/1186658/114666582-b52a3300-9cfe-11eb-9da1-9ccb539f54f7.png)

**Comment by kryczko** [comment]
I believe @rusty1s has already commented on these bugs in another issue, although I can't remember the issue itself. The tensorflow issue is fine, since pytorch is being used in this repo.

**Comment by thegodone** [comment]
Thanks kryczko but, I would appreciate to have a clear answer of Matthias especially because the last bug just pop up. It's always better to be sure than to speculate.

**Comment by rusty1s** [comment]
I think [this is the issue](https://github.com/rusty1s/pytorch_geometric/pull/1687) @kryczko has mentioned. Basically, we need to inherit the errors of `DimeNet(++)` if we want to make use of the pre-trained models, and want to provide the "fixed" version in case one wants to train a model from scratch.

**Comment by thegodone** [comment]
correct for the first issue, what about 2 and 5 ? that would be nice to have a switch to use "alpha-pretrain-dimenet" or "fix-dimenet" as you said.

**Comment by rusty1s** [comment]
Yes, we can automatically set this variable in case the user makes use of `from_qm9_pretrained` or not.

**Comment by thegodone** [comment]
my changes to get pp blocks:
`class InteractionBlockPP(torch.nn.Module):
    def __init__(self, hidden_channels, int_emb_size, basis_emb_size, num_spherical,
                 num_radial, num_before_skip, num_after_skip, act=swish):
        super(InteractionBlockPP, self).__init__()

        self.act = act
        self.basis_emb_size = basis_emb_size
        self.hidden_channels = hidden_channels
        self.num_spherical = num_spherical
        self.num_radial = num_radial
        
        # two layers 
        self.lin_rbf = Seq(Linear(num_radial, basis_emb_size, bias=False), Linear(basis_emb_size, hidden_channels, bias=False))
        
        self.lin_sbf = Seq(Linear(num_spherical * num_radial, basis_emb_size ,bias=False),Linear(basis_emb_size, int_emb_size,bias=False)) 
            
        self.xdownproj = Linear(hidden_channels, int_emb_size)
        self.xupproj = Linear(int_emb_size, hidden_channels)
        
        # Dense transformations of input messages.
        self.lin_kj = Linear(hidden_channels, hidden_channels)
        self.lin_ji = Linear(hidden_channels, hidden_channels)
        
        self.layers_before_skip = torch.nn.ModuleList([
            ResidualLayer(hidden_channels, act) for _ in range(num_before_skip)
        ])
        self.lin = Linear(hidden_channels, hidden_channels)
        self.layers_after_skip = torch.nn.ModuleList([
            ResidualLayer(hidden_channels, act) for _ in range(num_after_skip)
        ])

        self.reset_parameters()

    def reset_parameters(self):
        glorot_orthogonal(self.xdownproj.weight, scale=2.0)
        glorot_orthogonal(self.xupproj.weight, scale=2.0)
        glorot_orthogonal(self.lin_kj.weight, scale=2.0)
        self.lin_kj.bias.data.fill_(0)
        glorot_orthogonal(self.lin_ji.weight, scale=2.0)
        self.lin_ji.bias.data.fill_(0)
        #self.W.data.normal_(mean=0, std=2 / self.W.size(0))
        for res_layer in self.layers_before_skip:
            res_layer.reset_parameters()
        glorot_orthogonal(self.lin.weight, scale=2.0)
        self.lin.bias.data.fill_(0)
        for res_layer in self.layers_after_skip:
            res_layer.reset_parameters()

    def forward(self, x, rbf, sbf, idx_kj, idx_ji):
        rbf = self.lin_rbf(rbf)
        sbf = self.lin_sbf(sbf) 
        
        # Bessel basis transformation
        x_ji = self.act(self.lin_ji(x))
        x_kj = self.act(self.lin_kj(x))
        x_kj = x_kj * rbf

        # downproject
        x_kj = self.xdownproj(x_kj) # downproject to 32 as suggested, not sure if needed for scatter (instead they have tensorflow with gather)
        
        ## missing piece here x_kj = tf.gather(x_kj, id_expand_kj)
        # x_kj = torch.gather(x_kj, idx_kj) # added in the recent code Dimenet++
        
        # Spherical basis transformation
        # this look like identical to gather is it faster ???
        x_kj = x_kj[idx_kj] * sbf # hadamard product

        # aggregate interactions and up-project embeddings
        x_kj = scatter(x_kj, idx_ji, dim=0, dim_size=x.size(0))
        
        x_kj = self.xupproj(x_kj) # up project to 128 as suggested
        
        # transformations before skip connection            
        h = x_ji + x_kj
        for layer in self.layers_before_skip:
            h = layer(h)
            
        # transform before skip connection
        h = self.act(self.lin(h)) + x
        
        # transformation after skip connection
        for layer in self.layers_after_skip:
            h = layer(h)

        return h    
    
class OutputBlockPP(torch.nn.Module):
    def __init__(self, num_radial, hidden_channels, out_emb_size, out_channels, num_layers,
                 act=swish):
        super(OutputBlockPP, self).__init__()
        self.act = act

        self.lin_rbf = Linear(num_radial, hidden_channels, bias=False)
        self.xupproj = Linear(hidden_channels, out_emb_size, bias=False)

        self.lins = torch.nn.ModuleList()
        for _ in range(num_layers):
            self.lins.append(Linear(out_emb_size, out_emb_size))
        self.lin = Linear(out_emb_size, out_channels, bias=False)

        self.reset_parameters()

    def reset_parameters(self):
        glorot_orthogonal(self.lin_rbf.weight, scale=2.0)
        for lin in self.lins:
            glorot_orthogonal(lin.weight, scale=2.0)
            lin.bias.data.fill_(0)
        self.lin.weight.data.fill_(0)

    def forward(self, x, rbf, i, num_nodes=None):
        x = self.lin_rbf(rbf) * x #  g = self.dense_rbf(rbf) , x = g * x
        x = scatter(x, i, dim=0, dim_size=num_nodes) # x = tf.math.unsorted_segment_sum(x, idnb_i, n_atoms)
        
        x = self.xupproj(x)
        # this imply that the lin have to change to
        for lin in self.lins:
            x = self.act(lin(x)) 
        # return final dense layer done
        return self.lin(x)`

**Comment by rusty1s** [comment]
@thegodone Thanks! Can you send your implementation as a PR?

**Comment by thegodone** [comment]
I am working on the PR but got an issue #2413
